src/otx/algo/classification/metrics/hlabel_accuracy.py

```diff
@@ -0,0 +1,117 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Module for defining hierarchical label accuracy metric."""
+
+from __future__ import annotations
+
+import torch
+from torchmetrics import Metric
+from torchmetrics.classification import Accuracy, MultilabelAccuracy
+
+
+class HLabelAccuracy(Metric):
+    """Custom accuracy metric for h-label classification.
+
+    Args:
+        num_multiclass_heads (int): Number of multi-class heads.
+        num_multilabel_classes (int): Number of multi-label classes.
+        threshold_multilabel (float): Predictions with scores under the thresholds
+                                        are considered as negative. Defaults to 0.5.
+        head_idx_to_logits_range (dict[str, tuple[int, int]]): The range of logits which represents
+                                                                the number of classes for each heads.
+    """
+
+    def __init__(
+        self,
+        num_multiclass_heads: int,
+        num_multilabel_classes: int,
+        threshold_multilabel: float = 0.5,
+        head_idx_to_logits_range: dict[str, tuple[int, int]] | None = None,
+    ):
+        super().__init__()
+
+        self.num_multiclass_heads = num_multiclass_heads
+        self.num_multilabel_classes = num_multilabel_classes
+        self.threshold_multilabel = threshold_multilabel
+        self.head_idx_to_logits_range = head_idx_to_logits_range
+
+        # Multiclass classification accuracy will be defined later
+        self.multiclass_head_accuracy: list[Accuracy] | None = None
+
+        # Multilabel classification accuracy metrics
+        if num_multilabel_classes > 0:
+            # self.multilabel_accuracy = MultilabelAveragePrecision(
+            #     num_labels=self.num_multilabel_classes, average='macro'
+            # )
+            self.multilabel_accuracy = MultilabelAccuracy(
+                num_labels=self.num_multilabel_classes,
+                threshold=0.5,
+                average="macro",
+            )
+
+        if num_multiclass_heads == 0:
+            msg = "The number of multiclass heads should be larger than 0"
+            raise ValueError(msg)
+
+    def _set_multiclass_head_accuracy(self, device: str) -> None:
+        """Set the multiclass metric for each multiclass heads."""
+        if isinstance(self.head_idx_to_logits_range, dict):
+            self.multiclass_head_accuracy = [
+                Accuracy(
+                    task="multiclass",
+                    num_classes=int(head_range[1] - head_range[0]),
+                ).to(device)
+                for head_range in self.head_idx_to_logits_range.values()
+            ]
+        else:
+            msg = "head_idx_to_logits_range should be set."
+            raise TypeError(msg)
+
+    def update(self, preds: torch.Tensor, target: torch.Tensor) -> None:
+        """Update state with predictions and targets."""
+        if not self.multiclass_head_accuracy:
+            self._set_multiclass_head_accuracy(preds.device)
```

Comment:
Why should this be created at the update time?

Reply:
In the previous implementation, this was needed because there is no h-label related information at the initialized phase. 
However, by using the `HLabelMetaInfo` made by `DataModule`, I removed this logic. https://github.com/openvinotoolkit/training_extensions/pull/2761/commits/fdc033ec6b1fdc6615705ddd75af960a49af9f1d